3DGenZ/genz3d/kpconv/results/Log_2020-09-15_16-40-46/parameters.txt
# -----------------------------------#
# Parameters of the training session #
# -----------------------------------#

# Input parameters
# ****************

dataset = SemanticKitti
dataset_task = slam_segmentation
num_classes = 20
in_points_dim = 3
in_features_dim = 2
in_radius = 4.000000
input_threads = 10

# Model parameters
# ****************

architecture = simple resnetb resnetb_strided resnetb resnetb resnetb_strided resnetb resnetb resnetb_strided resnetb resnetb resnetb_strided resnetb nearest_upsample unary nearest_upsample unary nearest_upsample unary nearest_upsample unary
equivar_mode = 
invar_mode = 
num_layers = 5
first_features_dim = 128
use_batch_norm = 1
batch_norm_momentum = 0.020000

segmentation_ratio = 1.000000

# KPConv parameters
# *****************

first_subsampling_dl = 0.060000
num_kernel_points = 15
conv_radius = 2.500000
deform_radius = 6.000000
fixed_kernel_points = center
KP_extent = 1.200000
KP_influence = linear
aggregation_mode = sum
modulated = 0
n_frames = 1
max_in_points = 5768

max_val_points = 7823

val_radius = 4.000000

# Training parameters
# *******************

learning_rate = 0.010000
momentum = 0.980000
lr_decay_epochs = 1:0.984767 2:0.984767 3:0.984767 4:0.984767 5:0.984767 6:0.984767 7:0.984767 8:0.984767 9:0.984767 10:0.984767 11:0.984767 12:0.984767 13:0.984767 14:0.984767 15:0.984767 16:0.984767 17:0.984767 18:0.984767 19:0.984767 20:0.984767 21:0.984767 22:0.984767 23:0.984767 24:0.984767 25:0.984767 26:0.984767 27:0.984767 28:0.984767 29:0.984767 30:0.984767 31:0.984767 32:0.984767 33:0.984767 34:0.984767 35:0.984767 36:0.984767 37:0.984767 38:0.984767 39:0.984767 40:0.984767 41:0.984767 42:0.984767 43:0.984767 44:0.984767 45:0.984767 46:0.984767 47:0.984767 48:0.984767 49:0.984767 50:0.984767 51:0.984767 52:0.984767 53:0.984767 54:0.984767 55:0.984767 56:0.984767 57:0.984767 58:0.984767 59:0.984767 60:0.984767 61:0.984767 62:0.984767 63:0.984767 64:0.984767 65:0.984767 66:0.984767 67:0.984767 68:0.984767 69:0.984767 70:0.984767 71:0.984767 72:0.984767 73:0.984767 74:0.984767 75:0.984767 76:0.984767 77:0.984767 78:0.984767 79:0.984767 80:0.984767 81:0.984767 82:0.984767 83:0.984767 84:0.984767 85:0.984767 86:0.984767 87:0.984767 88:0.984767 89:0.984767 90:0.984767 91:0.984767 92:0.984767 93:0.984767 94:0.984767 95:0.984767 96:0.984767 97:0.984767 98:0.984767 99:0.984767 100:0.984767 101:0.984767 102:0.984767 103:0.984767 104:0.984767 105:0.984767 106:0.984767 107:0.984767 108:0.984767 109:0.984767 110:0.984767 111:0.984767 112:0.984767 113:0.984767 114:0.984767 115:0.984767 116:0.984767 117:0.984767 118:0.984767 119:0.984767 120:0.984767 121:0.984767 122:0.984767 123:0.984767 124:0.984767 125:0.984767 126:0.984767 127:0.984767 128:0.984767 129:0.984767 130:0.984767 131:0.984767 132:0.984767 133:0.984767 134:0.984767 135:0.984767 136:0.984767 137:0.984767 138:0.984767 139:0.984767 140:0.984767 141:0.984767 142:0.984767 143:0.984767 144:0.984767 145:0.984767 146:0.984767 147:0.984767 148:0.984767 149:0.984767 150:0.984767 151:0.984767 152:0.984767 153:0.984767 154:0.984767 155:0.984767 156:0.984767 157:0.984767 158:0.984767 159:0.984767 160:0.984767 161:0.984767 162:0.984767 163:0.984767 164:0.984767 165:0.984767 166:0.984767 167:0.984767 168:0.984767 169:0.984767 170:0.984767 171:0.984767 172:0.984767 173:0.984767 174:0.984767 175:0.984767 176:0.984767 177:0.984767 178:0.984767 179:0.984767 180:0.984767 181:0.984767 182:0.984767 183:0.984767 184:0.984767 185:0.984767 186:0.984767 187:0.984767 188:0.984767 189:0.984767 190:0.984767 191:0.984767 192:0.984767 193:0.984767 194:0.984767 195:0.984767 196:0.984767 197:0.984767 198:0.984767 199:0.984767 200:0.984767 201:0.984767 202:0.984767 203:0.984767 204:0.984767 205:0.984767 206:0.984767 207:0.984767 208:0.984767 209:0.984767 210:0.984767 211:0.984767 212:0.984767 213:0.984767 214:0.984767 215:0.984767 216:0.984767 217:0.984767 218:0.984767 219:0.984767 220:0.984767 221:0.984767 222:0.984767 223:0.984767 224:0.984767 225:0.984767 226:0.984767 227:0.984767 228:0.984767 229:0.984767 230:0.984767 231:0.984767 232:0.984767 233:0.984767 234:0.984767 235:0.984767 236:0.984767 237:0.984767 238:0.984767 239:0.984767 240:0.984767 241:0.984767 242:0.984767 243:0.984767 244:0.984767 245:0.984767 246:0.984767 247:0.984767 248:0.984767 249:0.984767 250:0.984767
grad_clip_norm = 100.000000

augment_symmetries = 1 0 0
augment_rotation = vertical
augment_noise = 0.001000
augment_occlusion = none
augment_occlusion_ratio = 0.200000
augment_occlusion_num = 1
augment_scale_anisotropic = 1
augment_scale_min = 0.800000
augment_scale_max = 1.200000
augment_color = 0.800000

weight_decay = 0.001000
segloss_balance = none
class_w =
deform_fitting_mode = point2point
deform_fitting_power = 1.000000
deform_lr_factor = 0.100000
repulse_extent = 1.200000
batch_num = 8
val_batch_num = 8
max_epoch = 800
epoch_steps = 500
validation_size = 200
checkpoint_gap = 50


3DGenZ/genz3d/kpconv/results/Log_2020-09-18_19-00-03/parameters.txt
# -----------------------------------#
# Parameters of the training session #
# -----------------------------------#

# Input parameters
# ****************

dataset = SemanticKitti
dataset_task = slam_segmentation
num_classes = 20
in_points_dim = 3
in_features_dim = 2
in_radius = 4.000000
input_threads = 1

# Model parameters
# ****************

architecture = simple resnetb resnetb_strided resnetb resnetb resnetb_strided resnetb resnetb resnetb_strided resnetb resnetb resnetb_strided resnetb nearest_upsample unary nearest_upsample unary nearest_upsample unary nearest_upsample unary
equivar_mode = 
invar_mode = 
num_layers = 5
first_features_dim = 128
use_batch_norm = 1
batch_norm_momentum = 0.020000

segmentation_ratio = 1.000000

# KPConv parameters
# *****************

first_subsampling_dl = 0.060000
num_kernel_points = 15
conv_radius = 2.500000
deform_radius = 6.000000
fixed_kernel_points = center
KP_extent = 1.200000
KP_influence = linear
aggregation_mode = sum
modulated = 0
n_frames = 1
max_in_points = 5768

max_val_points = 7823

val_radius = 4.000000

# Training parameters
# *******************

learning_rate = 0.010000
momentum = 0.980000
lr_decay_epochs = 1:0.984767 2:0.984767 3:0.984767 4:0.984767 5:0.984767 6:0.984767 7:0.984767 8:0.984767 9:0.984767 10:0.984767 11:0.984767 12:0.984767 13:0.984767 14:0.984767 15:0.984767 16:0.984767 17:0.984767 18:0.984767 19:0.984767 20:0.984767 21:0.984767 22:0.984767 23:0.984767 24:0.984767 25:0.984767 26:0.984767 27:0.984767 28:0.984767 29:0.984767 30:0.984767 31:0.984767 32:0.984767 33:0.984767 34:0.984767 35:0.984767 36:0.984767 37:0.984767 38:0.984767 39:0.984767 40:0.984767 41:0.984767 42:0.984767 43:0.984767 44:0.984767 45:0.984767 46:0.984767 47:0.984767 48:0.984767 49:0.984767 50:0.984767 51:0.984767 52:0.984767 53:0.984767 54:0.984767 55:0.984767 56:0.984767 57:0.984767 58:0.984767 59:0.984767 60:0.984767 61:0.984767 62:0.984767 63:0.984767 64:0.984767 65:0.984767 66:0.984767 67:0.984767 68:0.984767 69:0.984767 70:0.984767 71:0.984767 72:0.984767 73:0.984767 74:0.984767 75:0.984767 76:0.984767 77:0.984767 78:0.984767 79:0.984767 80:0.984767 81:0.984767 82:0.984767 83:0.984767 84:0.984767 85:0.984767 86:0.984767 87:0.984767 88:0.984767 89:0.984767 90:0.984767 91:0.984767 92:0.984767 93:0.984767 94:0.984767 95:0.984767 96:0.984767 97:0.984767 98:0.984767 99:0.984767 100:0.984767 101:0.984767 102:0.984767 103:0.984767 104:0.984767 105:0.984767 106:0.984767 107:0.984767 108:0.984767 109:0.984767 110:0.984767 111:0.984767 112:0.984767 113:0.984767 114:0.984767 115:0.984767 116:0.984767 117:0.984767 118:0.984767 119:0.984767 120:0.984767 121:0.984767 122:0.984767 123:0.984767 124:0.984767 125:0.984767 126:0.984767 127:0.984767 128:0.984767 129:0.984767 130:0.984767 131:0.984767 132:0.984767 133:0.984767 134:0.984767 135:0.984767 136:0.984767 137:0.984767 138:0.984767 139:0.984767 140:0.984767 141:0.984767 142:0.984767 143:0.984767 144:0.984767 145:0.984767 146:0.984767 147:0.984767 148:0.984767 149:0.984767 150:0.984767 151:0.984767 152:0.984767 153:0.984767 154:0.984767 155:0.984767 156:0.984767 157:0.984767 158:0.984767 159:0.984767 160:0.984767 161:0.984767 162:0.984767 163:0.984767 164:0.984767 165:0.984767 166:0.984767 167:0.984767 168:0.984767 169:0.984767 170:0.984767 171:0.984767 172:0.984767 173:0.984767 174:0.984767 175:0.984767 176:0.984767 177:0.984767 178:0.984767 179:0.984767 180:0.984767 181:0.984767 182:0.984767 183:0.984767 184:0.984767 185:0.984767 186:0.984767 187:0.984767 188:0.984767 189:0.984767 190:0.984767 191:0.984767 192:0.984767 193:0.984767 194:0.984767 195:0.984767 196:0.984767 197:0.984767 198:0.984767 199:0.984767 200:0.984767 201:0.984767 202:0.984767 203:0.984767 204:0.984767 205:0.984767 206:0.984767 207:0.984767 208:0.984767 209:0.984767 210:0.984767 211:0.984767 212:0.984767 213:0.984767 214:0.984767 215:0.984767 216:0.984767 217:0.984767 218:0.984767 219:0.984767 220:0.984767 221:0.984767 222:0.984767 223:0.984767 224:0.984767 225:0.984767 226:0.984767 227:0.984767 228:0.984767 229:0.984767 230:0.984767 231:0.984767 232:0.984767 233:0.984767 234:0.984767 235:0.984767 236:0.984767 237:0.984767 238:0.984767 239:0.984767 240:0.984767 241:0.984767 242:0.984767 243:0.984767 244:0.984767 245:0.984767 246:0.984767 247:0.984767 248:0.984767 249:0.984767 250:0.984767 
grad_clip_norm = 100.000000

augment_symmetries = 1 0 0
augment_rotation = vertical
augment_noise = 0.001000
augment_occlusion = none
augment_occlusion_ratio = 0.200000
augment_occlusion_num = 1
augment_scale_anisotropic = 1
augment_scale_min = 0.800000
augment_scale_max = 1.200000
augment_color = 0.800000

weight_decay = 0.001000
segloss_balance = none
class_w =
deform_fitting_mode = point2point
deform_fitting_power = 1.000000
deform_lr_factor = 0.100000
repulse_extent = 1.200000
batch_num = 8
val_batch_num = 8
max_epoch = 800
epoch_steps = 500
validation_size = 200
checkpoint_gap = 50


3DGenZ/genz3d/kpconv/utils/config.py
#
#
#      0=================================0
#      |    Kernel Point Convolutions    |
#      0=================================0
#
#
# ----------------------------------------------------------------------------------------------------------------------
#
#      Configuration class
#
# ----------------------------------------------------------------------------------------------------------------------
#
#      Hugues THOMAS - 11/06/2018
#


from os.path import join
import numpy as np


# Colors for printing
class bcolors:
    HEADER = '\033[95m'
    OKBLUE = '\033[94m'
    OKGREEN = '\033[92m'
    WARNING = '\033[93m'
    FAIL = '\033[91m'
    ENDC = '\033[0m'
    BOLD = '\033[1m'
    UNDERLINE = '\033[4m'


class Config:
    """
    Class containing the parameters you want to modify for this dataset
    """

    ##################
    # Input parameters
    ##################
    #list of ignore labels
    list_idx_ignore = [0, 3, 4, 7, 19]
    baseline = False
    convex_baseline = False 

    #ZSL parameter
    attribute_name = "glove_w2v"
    attribute_size = 600

    # Dataset name
    dataset = ''

    # Type of network model
    dataset_task = ''

    # Number of classes in the dataset
    num_classes = 0

    # Dimension of input points
    in_points_dim = 3

    # Dimension of input features
    in_features_dim = 1

    # Radius of the input sphere (ignored for models, only used for point clouds)
    in_radius = 1.0

    # Number of CPU threads for the input pipeline
    input_threads = 2#8

    ##################
    # Model parameters
    ##################

    # Architecture definition. List of blocks
    architecture = []

    # Decide the mode of equivariance and invariance
    equivar_mode = ''
    invar_mode = ''

    # Dimension of the first feature maps
    first_features_dim = 64

    # Batch normalization parameters
    use_batch_norm = True
    batch_norm_momentum = 0.99

    # For segmentation models : ratio between the segmented area and the input area
    segmentation_ratio = 1.0

    ###################
    # KPConv parameters
    ###################

    # Number of kernel points
    num_kernel_points = 15

    # Size of the first subsampling grid in meter
    first_subsampling_dl = 0.02

    # Radius of convolution in "number grid cell". (2.5 is the standard value)
    conv_radius = 2.5

    # Radius of deformable convolution in "number grid cell". Larger so that deformed kernel can spread out
    deform_radius = 5.0

    # Kernel point influence radius
    KP_extent = 1.0

    # Influence function when d < KP_extent. ('constant', 'linear', 'gaussian') When d > KP_extent, always zero
    KP_influence = 'linear'

    # Aggregation function of KPConv in ('closest', 'sum')
    # Decide if you sum all kernel point influences, or if you only take the influence of the closest KP
    aggregation_mode = 'sum'

    # Fixed points in the kernel : 'none', 'center' or 'verticals'
    fixed_kernel_points = 'center'

    # Use modulateion in deformable convolutions
    modulated = False

    # For SLAM datasets like SemanticKitti number of frames used (minimum one)
    n_frames = 1

    # For SLAM datasets like SemanticKitti max number of point in input cloud + validation
    max_in_points = 0
    val_radius = 51.0
    max_val_points = 50000

    #####################
    # Training parameters
    #####################

    # Network optimizer parameters (learning rate and momentum)
    learning_rate = 1e-3
    momentum = 0.9

    # Learning rate decays. Dictionary of all decay values with their epoch {epoch: decay}.
    lr_decays = {200: 0.2, 300: 0.2}

    # Gradient clipping value (negative means no clipping)
    grad_clip_norm = 100.0

    # Augmentation parameters
    augment_scale_anisotropic = True
    augment_scale_min = 0.9
    augment_scale_max = 1.1
    augment_symmetries = [False, False, False]
    augment_rotation = 'vertical'
    augment_noise = 0.005
    augment_color = 0.7

    # Augment with occlusions (not implemented yet)
    augment_occlusion = 'none'
    augment_occlusion_ratio = 0.2
    augment_occlusion_num = 1

    # Regularization loss importance
    weight_decay = 1e-3

    # The way we balance segmentation loss DEPRECATED
    segloss_balance = 'none'

    # Choose weights for class (used in segmentation loss). Empty list for no weights
    class_w = []

    # Deformable offset loss
    # 'point2point' fitting geometry by penalizing distance from deform point to input points
    # 'point2plane' fitting geometry by penalizing distance from deform point to input point triplet (not implemented)
    deform_fitting_mode = 'point2point'
    deform_fitting_power = 1.0              # Multiplier for the fitting/repulsive loss
    deform_lr_factor = 0.1                  # Multiplier for learning rate applied to the deformations
    repulse_extent = 1.0                    # Distance of repulsion for deformed kernel points

    # Number of batch
    batch_num = 10
    val_batch_num = 10

    # Maximal number of epochs
    max_epoch = 1000

    # Number of steps per epochs
    epoch_steps = 1000

    # Number of validation examples per epoch
    validation_size = 100

    # Number of epoch between each checkpoint
    checkpoint_gap = 50

    # Do we nee to save convergence
    saving = True
    saving_path = None

    def __init__(self):
        """
        Class Initialyser
        """

        # Number of layers
        self.num_layers = len([block for block in self.architecture if 'pool' in block or 'strided' in block]) + 1

        ###################
        # Deform layer list
        ###################
        #
        # List of boolean indicating which layer has a deformable convolution
        #

        layer_blocks = []
        self.deform_layers = []
        arch = self.architecture
        for block_i, block in enumerate(arch):

            # Get all blocks of the layer
            if not ('pool' in block or 'strided' in block or 'global' in block or 'upsample' in block):
                layer_blocks += [block]
                continue

            # Convolution neighbors indices
            # *****************************

            deform_layer = False
            if layer_blocks:
                if np.any(['deformable' in blck for blck in layer_blocks]):
                    deform_layer = True

            if 'pool' in block or 'strided' in block:
                if 'deformable' in block:
                    deform_layer = True

            self.deform_layers += [deform_layer]
            layer_blocks = []

            # Stop when meeting a global pooling or upsampling
            if 'global' in block or 'upsample' in block:
                break

    def load(self, path):

        filename = join(path, 'parameters.txt')
        with open(filename, 'r') as f:
            lines = f.readlines()

        # Class variable dictionary
        for line in lines:
            line_info = line.split()
            if len(line_info) > 2 and line_info[0] != '#':

                if line_info[2] == 'None':
                    setattr(self, line_info[0], None)

                elif line_info[0] == 'lr_decay_epochs':
                    self.lr_decays = {int(b.split(':')[0]): float(b.split(':')[1]) for b in line_info[2:]}

                elif line_info[0] == 'architecture':
                    self.architecture = [b for b in line_info[2:]]

                elif line_info[0] == 'augment_symmetries':
                    self.augment_symmetries = [bool(int(b)) for b in line_info[2:]]

                elif line_info[0] == 'num_classes':
                    if len(line_info) > 3:
                        self.num_classes = [int(c) for c in line_info[2:]]
                    else:
                        self.num_classes = int(line_info[2])

                elif line_info[0] == 'class_w':
                    self.class_w = [float(w) for w in line_info[2:]]

                elif hasattr(self, line_info[0]):
                    attr_type = type(getattr(self, line_info[0]))
                    if attr_type == bool:
                        setattr(self, line_info[0], attr_type(int(line_info[2])))
                    else:
                        setattr(self, line_info[0], attr_type(line_info[2]))

        self.saving = True
        self.saving_path = path
        self.__init__()

    def save(self):

        with open(join(self.saving_path, 'parameters.txt'), "w") as text_file:

            text_file.write('# -----------------------------------#\n')
            text_file.write('# Parameters of the training session #\n')
            text_file.write('# -----------------------------------#\n\n')
            #My own parameters
            text_file.write('# -----------------------------------#\n')
            text_file.write('# ZSL parameter #\n')
            text_file.write('# -----------------------------------#\n\n')
            
            
            text_file.write('# list_idx_ignore = {}\n'.format(self.list_idx_ignore))
            text_file.write('# baseline = {}\n'.format(self.baseline))
            text_file.write('# convex_baseline = {}\n'.format(self.convex_baseline))
            text_file.write('# attribute_name = {}\n'.format(self.attribute_name))
            text_file.write('# attribute_size = {}\n'.format(self.attribute_size))


            # Input parameters
            text_file.write('# Input parameters\n')
            text_file.write('# ****************\n\n')
            text_file.write('dataset = {:s}\n'.format(self.dataset))
            text_file.write('dataset_task = {:s}\n'.format(self.dataset_task))
            if type(self.num_classes) is list:
                text_file.write('num_classes =')
                for n in self.num_classes:
                    text_file.write(' {:d}'.format(n))
                text_file.write('\n')
            else:
                text_file.write('num_classes = {:d}\n'.format(self.num_classes))
            text_file.write('in_points_dim = {:d}\n'.format(self.in_points_dim))
            text_file.write('in_features_dim = {:d}\n'.format(self.in_features_dim))
            text_file.write('in_radius = {:.6f}\n'.format(self.in_radius))
            text_file.write('input_threads = {:d}\n\n'.format(self.input_threads))

            # Model parameters
            text_file.write('# Model parameters\n')
            text_file.write('# ****************\n\n')

            text_file.write('architecture =')
            for a in self.architecture:
                text_file.write(' {:s}'.format(a))
            text_file.write('\n')
            text_file.write('equivar_mode = {:s}\n'.format(self.equivar_mode))
            text_file.write('invar_mode = {:s}\n'.format(self.invar_mode))
            text_file.write('num_layers = {:d}\n'.format(self.num_layers))
            text_file.write('first_features_dim = {:d}\n'.format(self.first_features_dim))
            text_file.write('use_batch_norm = {:d}\n'.format(int(self.use_batch_norm)))
            text_file.write('batch_norm_momentum = {:.6f}\n\n'.format(self.batch_norm_momentum))
            text_file.write('segmentation_ratio = {:.6f}\n\n'.format(self.segmentation_ratio))

            # KPConv parameters
            text_file.write('# KPConv parameters\n')
            text_file.write('# *****************\n\n')

            text_file.write('first_subsampling_dl = {:.6f}\n'.format(self.first_subsampling_dl))
            text_file.write('num_kernel_points = {:d}\n'.format(self.num_kernel_points))
            text_file.write('conv_radius = {:.6f}\n'.format(self.conv_radius))
            text_file.write('deform_radius = {:.6f}\n'.format(self.deform_radius))
            text_file.write('fixed_kernel_points = {:s}\n'.format(self.fixed_kernel_points))
            text_file.write('KP_extent = {:.6f}\n'.format(self.KP_extent))
            text_file.write('KP_influence = {:s}\n'.format(self.KP_influence))
            text_file.write('aggregation_mode = {:s}\n'.format(self.aggregation_mode))
            text_file.write('modulated = {:d}\n'.format(int(self.modulated)))
            text_file.write('n_frames = {:d}\n'.format(self.n_frames))
            text_file.write('max_in_points = {:d}\n\n'.format(self.max_in_points))
            text_file.write('max_val_points = {:d}\n\n'.format(self.max_val_points))
            text_file.write('val_radius = {:.6f}\n\n'.format(self.val_radius))

            # Training parameters
            text_file.write('# Training parameters\n')
            text_file.write('# *******************\n\n')

            text_file.write('learning_rate = {:f}\n'.format(self.learning_rate))
            text_file.write('momentum = {:f}\n'.format(self.momentum))
            text_file.write('lr_decay_epochs =')
            for e, d in self.lr_decays.items():
                text_file.write(' {:d}:{:f}'.format(e, d))
            text_file.write('\n')
            text_file.write('grad_clip_norm = {:f}\n\n'.format(self.grad_clip_norm))


            text_file.write('augment_symmetries =')
            for a in self.augment_symmetries:
                text_file.write(' {:d}'.format(int(a)))
            text_file.write('\n')
            text_file.write('augment_rotation = {:s}\n'.format(self.augment_rotation))
            text_file.write('augment_noise = {:f}\n'.format(self.augment_noise))
            text_file.write('augment_occlusion = {:s}\n'.format(self.augment_occlusion))
            text_file.write('augment_occlusion_ratio = {:.6f}\n'.format(self.augment_occlusion_ratio))
            text_file.write('augment_occlusion_num = {:d}\n'.format(self.augment_occlusion_num))
            text_file.write('augment_scale_anisotropic = {:d}\n'.format(int(self.augment_scale_anisotropic)))
            text_file.write('augment_scale_min = {:.6f}\n'.format(self.augment_scale_min))
            text_file.write('augment_scale_max = {:.6f}\n'.format(self.augment_scale_max))
            text_file.write('augment_color = {:.6f}\n\n'.format(self.augment_color))

            text_file.write('weight_decay = {:f}\n'.format(self.weight_decay))
            text_file.write('segloss_balance = {:s}\n'.format(self.segloss_balance))
            text_file.write('class_w =')
            for a in self.class_w:
                text_file.write(' {:.6f}'.format(a))
            text_file.write('\n')
            text_file.write('deform_fitting_mode = {:s}\n'.format(self.deform_fitting_mode))
            text_file.write('deform_fitting_power = {:.6f}\n'.format(self.deform_fitting_power))
            text_file.write('deform_lr_factor = {:.6f}\n'.format(self.deform_lr_factor))
            text_file.write('repulse_extent = {:.6f}\n'.format(self.repulse_extent))
            text_file.write('batch_num = {:d}\n'.format(self.batch_num))
            text_file.write('val_batch_num = {:d}\n'.format(self.val_batch_num))
            text_file.write('max_epoch = {:d}\n'.format(self.max_epoch))
            if self.epoch_steps is None:
                text_file.write('epoch_steps = None\n')
            else:
                text_file.write('epoch_steps = {:d}\n'.format(self.epoch_steps))
            text_file.write('validation_size = {:d}\n'.format(self.validation_size))
            text_file.write('checkpoint_gap = {:d}\n'.format(self.checkpoint_gap))

